src/Compilers/CSharp/Test/Emit2/CodeGen/CodeGenSpanBasedStringConcatTests.cs

```diff
@@ -3232,86 +3263,144 @@ static void Main()
         // Instead of emitting this as a span-based concat of strings and chars we recognize "constantChar.ToString()" pattern and lower that arguments to constant strings
         verifier.VerifyIL("Test.M1", """
             {
-              // Code size       14 (0xe)
+              // Code size       34 (0x22)
               .maxstack  4
-              IL_0000:  ldstr      "c"
-              IL_0005:  ldarg.0
-              IL_0006:  ldarg.0
-              IL_0007:  ldarg.0
-              IL_0008:  call       "string string.Concat(string, string, string, string)"
-              IL_000d:  ret
+              .locals init (char V_0)
+              IL_0000:  ldc.i4.s   99
+              IL_0002:  stloc.0
+              IL_0003:  ldloca.s   V_0
+              IL_0005:  newobj     "System.ReadOnlySpan<char>..ctor(in char)"
+              IL_000a:  ldarg.0
+              IL_000b:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_0010:  ldarg.0
+              IL_0011:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_0016:  ldarg.0
+              IL_0017:  call       "System.ReadOnlySpan<char> string.op_Implicit(string)"
+              IL_001c:  call       "string string.Concat(System.ReadOnlySpan<char>, System.ReadOnlySpan<char>, System.ReadOnlySpan<char>, System.ReadOnlySpan<char>)"
+              IL_0021:  ret
             }
             """);
         verifier.VerifyIL("Test.M2", """
             {
-              // Code size       14 (0xe)
+              // Code size       34 (0x22)
```

Comment:
Here is a good concrete example (body of `M2` is `s + 'c'.ToString() + s + s`). Code size increased from 14 to 34, which is 2.4 times increase. At the same time benchmarks show that this not only didn't improve perf, but actually made it worse:
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22621.3007/22H2/2022Update/SunValley2)
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.200
  [Host]     : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.2 (8.0.224.6711), X64 RyuJIT AVX2


| Method        | Length | Mean      | Error    | StdDev   | Gen0   | Allocated |
|-------------- |------- |----------:|---------:|---------:|-------:|----------:|
| ConcatStrings | 1      |  11.13 ns | 0.073 ns | 0.065 ns | 0.0019 |      32 B |
| ConcatSpans   | 1      |  12.42 ns | 0.054 ns | 0.050 ns | 0.0019 |      32 B |
| ConcatStrings | 1000   | 178.33 ns | 3.058 ns | 2.860 ns | 0.3600 |    6024 B |
| ConcatSpans   | 1000   | 184.59 ns | 3.651 ns | 3.237 ns | 0.3600 |    6024 B |
```
```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Benchmarks>();

[MemoryDiagnoser]
public class Benchmarks
{
    private string _s;

    [Params(1, 1000)]
    public int Length { get; set; }

    [GlobalSetup]
    public void Setup() => _s = new string('s', Length);

    [Benchmark]
    public string ConcatStrings() => M1(_s);

    [Benchmark]
    public string ConcatSpans() => M2(_s);

    private static string M1(string s) => string.Concat(s, "c", s, s);

    private static string M2(string s)
    {
        var c = 'c';
        return string.Concat(s, new ReadOnlySpan<char>(in c), s, s);
    }
}
```
As you can see, span-based approach is slightly worse for short string length and the gap is getting larger as this length increases, so it doesn't seem like statistical noise and span codegen is clearly less efficient than string one. If we go back to the original issue the purpose of span-based concat was to avoid an intermediate allocation of `unknownChar.ToString()`, which slowed things down and added preasure to the GC. This is not a problem if we lower constant char to a constant string since constant string literals are interned by the runtime, so they are not allocated every time code is executed. Therefore it seems wasteful for me to emit more IL to get worse runtime performance and add unnecessary work to the JIT (since more IL corresponds to more time, required to compile it)